petastorm/namedtuple.py

```diff
@@ -0,0 +1,202 @@
+# Only Python 3.x needs a special namedtuple implementation for > 255 values.
+
+from __future__ import print_function
+
+import sys as _sys
+from collections import namedtuple
+from keyword import iskeyword as _iskeyword
+import six
+
+
+if six.PY2:
+    namedtuple_gt_255_fields = namedtuple
+else:
+    _class_template = """\
+from builtins import property as _property, tuple as _tuple
+from operator import itemgetter as _itemgetter
+from collections import OrderedDict
+
+class {typename}(tuple):
+    '{typename}({arg_list})'
+
+    __slots__ = ()
+
+    _fields = {field_names!r}
+
+    def __new__(_cls, *args, **kwargs):
+        'Create new instance of {typename}({arg_list})'
+        values = []
+        missings = []
+        for index, field in enumerate(_cls._fields):
+            if index < len(args):
+                values.append(args[index])
+                if field in kwargs:
+                    raise TypeError('__new__() got multiple values for argument %r' % field)
+            elif field in kwargs:
+                values.append(kwargs[field])
+            else:
+                values.append(None)
+                missings.append(field)
+
+        count_missings = len(missings)
+        if count_missings > 0:
+            last_missing = ' and %s' % missings[-1] if count_missings > 1 else ''
+            missings = missings[:-1] if count_missings > 1 else missings
+            raise TypeError('__new__() missing %d required positional argument%s: %s%s' %
+                            (count_missings, 's' if count_missings > 1 else '',
+                            ', '.join(missings), last_missing))
+
+        return _tuple.__new__(_cls, values)
+
+    @classmethod
+    def _make(cls, iterable, new=tuple.__new__, len=len):
+        'Make a new {typename} object from a sequence or iterable'
+        result = new(cls, iterable)
+        if len(result) != {num_fields:d}:
+            raise TypeError('Expected {num_fields:d} arguments, got %d' % len(result))
+        return result
+
+    def _replace(_self, **kwds):
+        'Return a new {typename} object replacing specified fields with new values'
+        result = _self._make(map(kwds.pop, {field_names!r}, _self))
+        if kwds:
+            raise ValueError('Got unexpected field names: %r' % list(kwds))
+        return result
+
+    def __repr__(self):
+        'Return a nicely formatted representation string'
+        return self.__class__.__name__ + '({repr_fmt})' % self
+
+    def _asdict(self):
+        'Return a new OrderedDict which maps field names to their values.'
+        return OrderedDict(zip(self._fields, self))
+
+    def __getnewargs__(self):
+        'Return self as a plain tuple. Used by copy and pickle.'
+        return tuple(self)
+
```

Comment:
We don't have `__dict__`. Is this on purpose?
https://github.com/python/cpython/blob/6cbb4c0795099b79f0a7c7d20df4ba1c1ec0ac24/Lib/collections.py#L290

Also some parts of the "def namedtuple_gt_255_fields" are different from python 2.7 (I mean apart from the  the differences related to the problem we are trying to solve). Don't see anything critical, but still interested in the reason of the differences.

Reply:
Yes, it makes __dict__ being the fields of the namedtuple with its values only.


The other small differences are because I based this work on another commit of collections.py (but ref it as master branch so...).

To make things clearer for the NOTICE, I can rebase it against the 2.7 implementation, that will make the NOTICE and the derived work changes more compact and focused only on the > 255 fields support.

Reply:
Not on purpose but it was not necessary in the end.

I will make a pass on the code in order to ensure it sticks as much as possible to the cython 2.7 implementation since the diffs are more than what is necessary.

Reply:
It comes from 3.6 implementation initially where the __dict__ is not redefined.